clients/src/main/java/org/apache/kafka/common/resource/Resource.java

```diff
@@ -22,14 +22,15 @@
 import java.util.Objects;
 
 /**
- * Represents a cluster resource with a tuple of (type, name).
+ * Represents a cluster resource with a tuple of (type, name, nameType).
```

Comment:
IMHO, no.

Conceptually, it doesn't make any sense for a _resource_ to support literal or prefixed names. A _resource_ should be an actual cluster resource, with a full name.

The two `Resource` classes are used in two different areas of the API: topic config and Acls.  For one area were being forced to add this `ResourceTypeName` to the `Resource` class. But in the other are the name type field is not required.

I would say that we leave these two classes as separate for now and look to merge them in KIP-50. This way, users wanting to make use of the topic config API don't need to worry about this new `ResourceNameType` field. This is only required if dealing with Acls, and only until we implement KIP-50, which should also make the field redundant.

Reply:
To be fair, `org.apache.kafka.common.requests.Resource` is not actually a part of the public AdminClient API for getting and setting configurations.  The public class for that is `org.apache.kafka.common.config.ConfigResource`.  I would argue that we should just get rid of `o.a.k.c.r.Resource`, since it just duplicates `ConfigResource`.  We just end up copying things to and from `ConfigResource` in `alterConfigs` and the other functions.  And it's an internal private class anyway, so we can always re-create it if we need to (we won't)

(I would argue that ConfigResource should never have been separate from `org.apache.kafka.common.resource.Resource`, but that ship has sailed since ConfigResource is part of the stable API now).

Reply:
By the way, I think we should remove `o.a.k.c.r.Resource` in a separate follow-on change.  It's not really related to KIP-290